pkg/logging/journald_writer.go

```diff
@@ -0,0 +1,122 @@
+type JournaldWriter struct {
+	fallbackWriter  io.Writer
+	fallbackEncoder zapcore.Encoder
+	loggingFatal    *atomic.Bool
+}
+
+// NewJournaldWriter returns a new JournaldWriter with the specified fallback encoder and writer,
+// which will be used when the log entry cannot be sent to systemd-journald.
+// loggingFatal is set to true when fatals are logged so that other non-fatals can be swallowed.
+// You can find an example usage under Logging.loggingFatal.
+func NewJournaldWriter(fallbackEncoder zapcore.Encoder, fallbackWriter io.Writer, loggingFatal *atomic.Bool) (io.Writer, error) {
+	return &JournaldWriter{
+		fallbackEncoder: fallbackEncoder,
+		fallbackWriter:  fallbackWriter,
+		loggingFatal:    loggingFatal,
+	}, nil
+}
+
+// Write sends the log entry to systemd-journald.
+// If the sending of the data to systemd-journald fails, the configured fallback writer is used.
+//
+// If fatal entries are logged, Write sleeps for a short time after the message has been written,
+// before zap then exits the program,
+// as systemd may not correctly attribute all or part of the log message due to a race condition,
+// and therefore it would be missing when checking logs via journalctl -u icingadb:
+// https://github.com/systemd/systemd/issues/2913
+//
+// Without sleep, we observed the following exit behavior when starting Icinga DB via systemd:
+//
+// • All or part of the log message is lost,
+// regardless of whether it is logged to stderr or to systemd-journald.
+//
+// • When writing to stderr after also writing to systemd-journald.
+// the message seems to be attributed correctly,
+// which is good for panics not under our control as they print to stderr.
+// For messages under our control,
+// this is actually not an option,
+// as we only want to log to systemd-journald.
+//
+// • Sleep before exiting helps systemd to attribute the message correctly,
+// regardless of whether it is written to systemd-journald or stderr.
+func (w *JournaldWriter) Write(p []byte) (int, error) {
+	var entry GobEntry
+	dec := gob.NewDecoder(bytes.NewBuffer(p))
+	if err := dec.Decode(&entry); err != nil {
+		return 0, errors.Wrapf(err, "can't decode %q", p)
+	}
+
+	if entry.Level == zapcore.FatalLevel {
+		// Sleep between fatal log and os.Exit() to overcome a race condition that
+		// results in systemd no longer attributing messages from processes
+		// that have exited to their cgroup.
+		// See: https://github.com/systemd/systemd/issues/2913.
+		defer time.Sleep(time.Millisecond * 1100)
+
+		w.loggingFatal.Store(true)
+	}
+
+	pri, ok := priorities[entry.Level]
+	if !ok {
+		return 0, errors.Errorf("unknown log level %q", entry.Level)
+	}
+
+	fields := map[string]string{
+		"SYSLOG_IDENTIFIER": filepath.Base(os.Args[0]),
+	}
+	enc := zapcore.NewMapObjectEncoder()
+	for i := range entry.Fields {
+		entry.Fields[i].AddTo(enc)
+	}
+	for k, v := range enc.Fields {
+		fields[utils.ConvertCamelCase(k, unicode.UpperCase, '_')] = fmt.Sprintf("%v", v)
```

Comment:
I think we should prefix the fields with `ICINGADB_` in oder to avoid collisions. Any thoughts on that?

Reply:
Given that it's just whatever you put into `logger.*w()`, sounds like a good idea.